src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -665,3 +674,120 @@ edgeSet = Set.fromAscList . edgeList
+-- part they belong to.
+--
+-- /Note/: as `AdjacencyMap` only represents __undirected__ bipartite graphs,
+-- all edges in the input graph are assumed to be bidirected and all edges in
+-- the output `AdjacencyMap` are bidirected.
+--
+-- It is advised to use 'leftVertexList' and 'rightVertexList' to obtain the
+-- partition of the vertices and 'hasLeftVertex' and 'hasRightVertex' to check
+-- whether a vertex belong to a part.
+--
+-- Complexity: /O((n + m) log(n))/.
+--
+-- @
+-- detectParts 'Algebra.Graph.AdjacencyMap.empty'                             == Just 'empty'
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.vertex' x))               == True
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.edge' x y))               == True
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.edges' [(1, 2), (1, 3)])) == True
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.clique' x))               == x < 3
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.star' x ys))              == not (x `elem` ys)
+-- 'Data.Maybe.isJust' (detectParts ('Algebra.Graph.AdjacencyMap.biclique' xs ys))         == True
+-- 'Data.Maybe.isJust' ('fromBipartite' ('toBipartite' am))       == True
+-- 'Data.Maybe.isJust' (detectParts (1 * 2 * 3))              == False
+-- 'Data.Maybe.isJust' (detectParts ((1 + 2) * (3 + 4)))      == True
+-- @
+detectParts :: Ord a => AM.AdjacencyMap a -> Maybe (AdjacencyMap a a)
+detectParts g = let s = AM.symmetricClosure g
+                 in build s <$> (foldM (runDfs s) Map.empty $ AM.vertexList s)
+    where
+        dfs :: Ord a => PartId -> AM.AdjacencyMap a -> Map.Map a PartId -> a -> Maybe (Map.Map a PartId)
+        dfs p g m v = foldM (action p g) (Map.insert v p m) $ neighbours v g
+
+        action :: Ord a => PartId -> AM.AdjacencyMap a -> Map.Map a PartId -> a -> Maybe (Map.Map a PartId)
+        action p g m v = case v `Map.lookup` m of
+                              Nothing -> dfs (otherPart p) g m v
+                              Just q  -> if q /= p then Just m else Nothing
+
+        runDfs :: Ord a => AM.AdjacencyMap a -> Map.Map a PartId -> a -> Maybe (Map.Map a PartId)
+        runDfs g m v = (m <$ Map.lookup v m) <|> dfs LeftPart g m v
+
+        neighbours :: Ord a => a -> AM.AdjacencyMap a -> Set.Set a
+        neighbours v g = fromJust $ v `Map.lookup` AM.adjacencyMap g
+
+        build :: Ord a => AM.AdjacencyMap a -> Map.Map a PartId -> AdjacencyMap a a
+        build g m = toBipartite $ AM.gmap (toEither m) g
+
+        toEither :: Ord a => Map.Map a PartId -> a -> Either a a
+        toEither m v = case fromJust (v `Map.lookup` m) of
+                            LeftPart  -> Left  v
+                            RightPart -> Right v
+
+
+type PartMap a = Map.Map a (PartId, NE.NonEmpty a)
+
+detectParts' :: Ord a => AM.AdjacencyMap a -> Either [a] (AdjacencyMap a a)
```

Comment:
Is this function as fast as `detectParts` in the `Right` case, i.e. if the graph is bipartite? If not and it is substantially slower than `detectParts`, then I think it's not worth combining the two tasks in one, and having a simpler `findOddCycle` would actually be better.

Reply:
Another question: Is there a more general and hopefully simpler function that we could write, so that both `findParts` and `findOddCycle` could be implemented by calling it?

Reply:
I don't know any linear algorithm for finding an odd cycle other than extending the bipartiteness testing algorithm in some way like here. What do you mean by "simplier `findOddCycle`?"

Reply:
> What do you mean by "simplier findOddCycle?"

I guess if you specialise this function to just finding the odd cycle, then it will become simpler. Or am I wrong? Are you saying that if the type signature were `findOddCycle :: Ord a => AM.AdjacencyMap a -> Maybe [a]` then the function would essentially remain the same?

Reply:
> the function would essentially remain the same

It really seems so to me. I'm going to have a try on making `findOddCycle` simpler anyway.

There are chances as the monadic idiom here is different. And for those usages of `Maybe` we already know that there are chances that DFS will look simpler.